### DISCUSSION TITLE
RFC: Submission APIs

### DISCUSSION THREAD

**Discussion by edmundhung** [edmundhung/conform#354] [Announcements]
## RFC: Submission APIs

Status: `Proposed`

Releated:
- #211
- #239

### Background

There are a few common questions when we parse the `FormData` on the server:

```tsx
export async function action({ request }: ActionArgs) {
  const formData = await request.formData();
  const submission = parse(formData, { schema });

  // 1. Why should we check the `submission.intent`? Can we skip it?
  if (!submission.value || submission.intent !== "submit") {
    return json(submission);
  }

  // 2. Can we display additional errors on the form if the job is failed?
  await save(submission.value);

  // 3. How do we reset the form if the submission is successful?
  return json(submission);
}

export default function Example() {
  const lastSubmission = useActionData<typeof action>();
  const [form, fields] = useForm({
    lastSubmission,
  });

  // 4. How do we know if the submission is failed?
}
```

For sure, we have an answer for all the questions above:

```tsx
export async function action({ request }: ActionArgs) {
  const formData = await request.formData();
  const submission = parse(formData, { schema });

  /**
   * Checking for `submission.intent` is necessary to avoid the server
   * from further processing data that is not ready. As Conform might call the server
   * for different purposes such as validation or list manipulation in cases like:
   * - No client validation
   * - Progressive enhancement (Before JS is loaded)
   * - Async validation
   */
  if (submission.intent !== "submit") {
    // Include a custom status to indicate that the submission is neither error or success
    return json({ status: "idle", submission } as const);
  }

  if (!submission.value) {
    // Send the `error`` status instead if no value is resolved
    return json({ status: "error", submission } as const, { status: 400 });
  }

  const result = await save(submission.value);

  if (!result.successful) {
    return json({
      status: "error",
      submission: {
        ...submission,
        error: {
          // Manipulate `submission.error` with additional messages
          ...submission.error,
          // The error will be populated on `form.errors`
          "": ["Submission failed"],
          // The error will be populated on `fields.address.errors`
          address: ["Address is invalid"],
        },
      },
    } as const);
  }

  // Conform will reset the form for you if the payload is `null`
  return json({
    status: "success",
    submission: { ...submission, payload: null },
  } as const);
}

export default function Example() {
  const actionData = useActionData<typeof action>();
  const [form, fields] = useForm({
    lastSubmission: actionData?.submission,
  });

  // Now we know the status of the submission with `actionData?.status`
}
```

However, the DX (Developer Experience) is poor. We can have all these documented, but it is still hard to understand what each of these code snippet does, especially for beginners. It could also be error-prone if you are not familiar with the data structure of the `submission` object.

We need proper APIs to handle the submission on the server.

### Proposal

Introducing an additional method to the `submission` object: `submission.reply()`.

```tsx
export async function action({ request }: ActionArgs) {
  const formData = await request.formData();
  const submission = parse(formData, { schema });

  /**
   * The submission status could be either "success", "error" or undefined
   * If the status is undefined, it means that the submission is not ready (i.e. `intent` is not `submit`)
   */
  if (submission.status !== 'success') {
    return json(submission.reply(), {
      // You can also use the status to determine the HTTP status code
      status: submission.status === 'error' ? 400 : 200,
    });
  }

  const result = await save(submission.value);

  if (!result.successful) {
    return json(
      submission.reply({
        // You can also pass additional error to the `reply` method
        formError: ["Submission failed"],
        fieldError: {
          address: ["Address is invalid"],
        },

        // or avoid sending the the field value back to client by specifying the field names
        hideFields: ['password'],
      })
    );
  } 

  // Reply the submission with `resetForm` option
  return json(submission.reply({ resetForm: true }));
}

export default function Example() {
  const lastResult = useActionData<typeof action>();
  const [form, fields] = useForm({
    // `lastSubmission` is renamed to `lastResult` to avoid confusion
    lastResult,
  });

  // We can now find out the status of the submission from the form metadata as well 
  console.log(form.status); // "success", "error" or undefined
}
```

**Comment by cevr** [comment]
This looks interesting!

I was thinking for hiding/augmenting submissions, maybe you can have two separate apis:
`reject/accept` for majority of use cases
`rejectWith/acceptWith` for augmentation

```tsx
// form schema is username & password
submission.rejectWith(sub => ({ username: sub.username })) // <-- hide password
```

**Comment by codyarose** [comment]
Zod strips out unrecognized keys during parsing by default, so [`.pick/.omit`](https://zod.dev/?id=pickomit) might be of use for avoiding sending certain field values back to the client.

Maybe `.reject/.accept` could take an optional `schema` property that would be used to revalidate the data (strip out omitted keys) before returning. You could use the same schema for parsing the submission and omitting properties you don't want to return to the client. 

```ts
const schema = z.object({
  username: z.string(),
  password: z.string(),
});

export async function action({ request }: ActionArgs) {
  const formData = await request.formData();
  const submission = parse(formData, { schema });

  if (!submission.value) {
    return json(submission.reject();
  }

  const result = await save(submission.value);
  
  if (!result.successful) {
    return json(
      submission.reject({
        formError: ["Submission failed"],
      })
    );
  }

  return json(submission.accept({ schema: schema.omit({ password: true }) }));
  // or .pick to explicitly define what you _do_ want to return
  // return json(submission.accept({ schema: schema.pick({ username: true }) }));
}
```

**Comment by DavoCg** [comment]
Hi @edmundhung 

I'm using : 

 `@conform-to/react": "^1.0.0-pre.5`
 `@conform-to/zod": "^1.0.0-pre.5`

I have a type error for the `fieldErrors` in `reply` : `Type 'null' is not assignable to type 'string[]'` 
I would expect being able to pass directly submission.error to fieldErrors if submission.error is defined. 

Is it something that is still subject to change ?

Thank a lot for all the work, v1 is very promising 

<img width="584" alt="Capture d’écran 2024-01-02 à 12 57 28" src="https://github.com/edmundhung/conform/assets/5452654/707a2458-79db-4d51-866e-c42eab0db0dc">

**Reply by edmundhung** [reply]
It will always reply with the submission error. The `formErrors` and `fieldErrors` options are useful mainly if you have some extra works that might failed and you wanna show additional error on the form. (e.g. failed to save the data in a database / failed to send the email)

**Reply by DavoCg** [reply]
Oh gotcha, I missed that, I'm indeed using this way for my action logic 

```javascript
if (submission.status === "success") {
    const payload = {
      email: submission.value.email,
      password: submission.value.password,
    };

    try {
      const headers = await signin(payload);
      const redirectUrl = await getRedirectUrl(request);
      return redirect(redirectUrl, { headers });
    } catch (err) {
      return json(
        submission.reply({
          formErrors: ["SIGNIN_ERROR"],
        })
      );
    }
  }
``` 

Thanks for the quick answer and happy new year

**Comment by dogbane-dev** [comment]
Hello, playing around with v1-pre5 and looking great! Had a question about the API

Should the submission API allow for the returning of some arbitrary value? For example the result of a mutation.

The type would have to specified as a generic argument to SubmissionResult but then I think could go something like this:
```
const result = db.user.update(...)
return submission.reply({
  data: result
})
```

So that we have access to it on the front end (via lastResult)

Or would this not be generally needed?

**Reply by edmundhung** [reply]
You are free to return anything on your action. Just making sure you pass the submission result to the `useForm` hook properly.

```tsx
export function action() {
  // ...
  const data = db.user.update(...)

  return {
    result: submission.reply(),
    data,
  };
}

export default function Example() {
   const actionData = useActionData();
   const [form, fields] = useForm({
     lastResult: actionData?.result,
   })

   // ...
}
```

**Reply by dogbane-dev** [reply]
Oh thank you, this didn't occur to me, I was stuck on needing to return `submission.reply()`

Appreciate it!

**Comment by dogbane-dev** [comment]
Also- this is a separate issue so I'll leave a separate comment but I am having some issues with the `form.dirty` property.

I am playing around with v1-pre5 with Next 14 and its been going largely very well, but am running into some unexpected behavior with the form dirty state.

So I have a form populated with some init values (requested on server at page.tsx level) and form.dirty correctly starts as `false`

What I want to happen is after a successful submission for form.dirty to go back to `false` - however it always seems to remains `true`. Even after a router.refresh, which gets the correct fresh default values, it remains dirty.

Using `submission.reply({ resetForm: true })` doesn't seem to help me either for whatever reason because the initial values go back to the pre-submission initial values.

Any ideas? Let me know if youd like me to make a minimal repro

Thanks

**Reply by edmundhung** [reply]
Oh, thanks a lot for testing this! Is it just the dirty flag stay `true`? Or is the form value remain unchanged as well?

A minimal repro will be awesome. You can use the nextjs example here: https://github.com/edmundhung/conform/tree/v1/examples/nextjs

**Reply by dogbane-dev** [reply]
I believe the form value is changing as expected but Im not positive I understand what Im looking at. I will work on a repro, thanks!

**Reply by dogbane-dev** [reply]
I made a repro based on the Todo example you provided (thank you): https://github.com/kwerdna19/conform-nextjs-repro

I only modified the TodoForm and created a psuedo-db to store the tasks/title on the server side.

**Reply by dogbane-dev** [reply]
edit: I thought I found a solution but I think it is not working actually after more playing around-- I will leave the solution below still:

Calling router.refresh to ensure the component has the updated data and then resetting the form seems to work.
```
  const status = form.status
  const router = useRouter()
  useEffect(() => {
    if(status === 'success') {
      router.refresh()
      form.dispatch(control.reset())
    }
  }, [status, form])
```

It feels a little icky but works as intended.

I also noticed that form.status never receives a success value unless you have `shouldValidate: 'onBlur'` in the useForm. This could be intended but worth noting

**Reply by edmundhung** [reply]
I notice a few things here:

- NextJS injects extra inputs ruins the dirty check logic. If you type something and then remove it, it will keep saying it is dirty because Conform find those `$ACTION_XXX` new when comparing with the default value. I will need to figure out some way to exclude it.
- I am not familiar with NextJS but it looks like the default value is not updated when server action returns the submission result.
- I kind of get it working with this:


```tsx
// actions.ts
export async function createTodos(prevState: unknown, formData: FormData) {
	const submission = parseWithZod(formData, {
		schema: todosSchema,
	});

	if (submission.status !== 'success') {
		return {
			result: submission.reply(),
		};
	}

        // At first I kept all the following lines uncommented.
        // But this only works if they are commented out......
        // Maybe a nextjs bug? 😅 
	// title = submission.value.title
	// tasks = []
	// submission.value.tasks.forEach(t => tasks.push(t))

	return {
		result: submission.reply({
			resetForm: true,
		}),
                // Send the updated default value back
		defaultValue: submission.value,
	}
}

// form.ts
export function TodoForm({ ... }) {
  const [data, action] = useFormState(createTodos, undefined);
  const [form, fields] = useForm({
    defaultValue: data?.defaultValue ?? defaultValue,
    lastResult: data?.result,
    onValidate({ formData }) {
      return parseWithZod(formData, { schema: todosSchema });
    },
    shouldValidate: 'onBlur',
  });

  // ...
}
```

**Reply by dogbane-dev** [reply]
Thanks for looking into it! I'll go ahead and try that tomorrow.

Another issue I noticed- not sure if intended, is moving the repeatable fields clears out any "unsaved" text.

It seems to be an issue caused by the inclusion of the `key` property on the input. If I omit it, the text no longer disappears when the field is moved.

Bug demo:


<video src="https://github.com/edmundhung/conform/assets/64341310/a61cf65e-fadb-4e03-ab44-31e8dc4d9770" />

**Reply by edmundhung** [reply]
I think the problem you see might be related to the useEffect approach you were testing? I don't see it happen to me while testing.

I spot one more problem in the implementation and get it working the way I expected now:
```tsx
// An in memory store
function createStore<T>(initialValue?: T) {
    let value = initialValue;

    return {
        getValue() {
            return value;
        },
        setValue(newValue: T) {
            value = newValue;
        }
    }
}

const todoStore = createStore<z.output<typeof todosSchema>>();

export async function getTodos() {
    return todoStore.getValue();
}

export async function createTodos(prevState: unknown, formData: FormData) {
    const submission = parseWithZod(formData, {
        schema: todosSchema,
    });

    if (submission.status !== 'success') {
        return submission.reply();
    }

    // Update the store value
    todoStore.setValue(submission.value);

    // Important: Without this the default value will never be changed
    revalidatePath('/todos')

    return submission.reply({
        resetForm: true,
    });
}
```

![NextJS - Conform Example](https://github.com/edmundhung/conform/assets/8258047/9d651dc3-ff33-4967-b55b-d5083c70d9af)

**Reply by dogbane-dev** [reply]
Thanks for the update- also I think the issue in the video I posted is unrelated- I am almost sure it is happening on an unmodified version of the example repo. Again- maybe it is expected behavior (?)

**Reply by edmundhung** [reply]
I have just released `1.0.0-pre.6`. If you wanna give it a try, just replace the action with the script I shared above and add a new option to the useForm hook like this:

```tsx
const [form, fields] = useForm({
    defaultValue,
    lastResult,
    onValidate({ formData }) {
        return parseWithZod(formData, { schema: todosSchema });
    },
    shouldValidate: 'onBlur',

    // New: this tells Conform to ignore those inputs generated by nextjs
    // when considering the dirty state 
    shouldDirtyConsider(name) {
        return !name.startsWith('$ACTION');
    },
});
```

Everything else should stay the same as you shared on the repo.

**Reply by dogbane-dev** [reply]
Thanks @edmundhung its starting to work as intended after submission! very cool thanks for the feature 👍 

One thing Im not so sure about- if you modify a field and then undo your change (so that is back to the default value) should dirty go back to false? Is this a bug?

<video src="https://github.com/edmundhung/conform/assets/64341310/1854f7ad-06e2-4f68-b09f-803625bd4de1"/>

**Reply by edmundhung** [reply]
It is exact why I added the `shouldDirtyConsider` option. Maybe you forgot to add it to the todos form?

**Reply by dogbane-dev** [reply]
Oh oops, you were so right. The example repo is working as expected now. I am still having problems with my personal project repo though for some reason despite having added `shouldDirtyConsider` for sure. Going to look into it more.

Internally what value is it comparing though ? Because from my console logs in my project it seems like form.value and the passed in defaultValue **are** deeply equal (aside from the $ACTION properties)

**Reply by dogbane-dev** [reply]
Ah I think I figured it out btw, I peered at the source code and saw you were using JSON.stringify to compare objects - mine was incorrectly returning dirty because the order of the fields was different . My default values were `[{ a: 'test', b: 'test' }]` whereas the form values had `[{ b: 'test', a: 'test' }]` - changing the order of the fields fixed it for me but I think this is a bug

**Reply by edmundhung** [reply]
This is definitely a bug 😅 I will fix that in the next pre-release.

**Comment by adaboese** [comment]
My grudge with the current design/proposal is that there is no way to communicate successful state message.

I just want a simple way to communicate form-level (as opposed to input level) success/failure of the submission, e.g.

```ts
return json(submission.reply({
  status: {
    type: 'success',
    message: 'Password reset link sent to your email'
  },
});
```

```ts
return json(submission.reply({
  status: {
    type: 'error',
    message: 'Incorrect email or password'
  },
});
```

At the moment, I have to add new top-level properties and it feels like I am working against the framework, e.g.

```ts
return json({
  ...submission.reply(),
  message: {
    message: 'Password reset email sent.',
    type: 'success' as const,
  },
});
```

**Reply by Pouet--** [reply]
Hello, I'm just quoting the example from @edmundhung , did you try to send the value back with the "formError" property ?

 ```tsx
 export async function action({ request }: ActionArgs) {
   const formData = await request.formData();
   const submission = parse(formData, { schema });
 
   [...]
 
   const result = await save(submission.value);
 
   if (!result.successful) {
     return json(
       submission.reply({
         // You can also pass additional error to the `reply` method
         formError: ["Submission failed"],
         fieldError: {
           address: ["Address is invalid"],
         },
 
         // or avoid sending the the field value back to client by specifying the field names
         hideFields: ['password'],
       })
     );
   } 
 
   // Reply the submission with `resetForm` option
   return json(submission.reply({ resetForm: true }));
 }
 ```

**Reply by adaboese** [reply]
You can with the `formError`, but then what about the success message?

**Reply by edmundhung** [reply]
There is nothing wrong from sending extra stuff beyond the submission result. This is also the pattern Kent did in the Epic Stack and how he taught in his workshops, like this:
```tsx
return json({
  // Always send the result in a named property
  result: submission.reply(),
  // So you can add more stuff later if you need
  message: {
    message: 'Password reset email sent.',
    type: 'success' as const,
  },
});
```

What would you suggest Conform to do for you if it accepts the success message?

**Reply by adaboese** [reply]
> What would you suggest Conform to do for you if it accepts the success message?

I would extend `submission.reply` to allow status message.

```ts
return json({
  result: submission.reply({
    status: {
      message: 'Password reset email sent.',
      type: 'success' as const,
    }
  }),
});
```

**Reply by edmundhung** [reply]
Yeah, you mentioned that already. How will you consume it in the route component? If you pass the message as I mentioned:

```tsx
export function action() {
  // ...

  return json({
    // Always send the result in a named property
    result: submission.reply(),
    // So you can add more stuff later if you need
    message: {
      message: 'Password reset email sent.',
      type: 'success' as const,
    },
  });
}

export default function Component() {
  const actionData = useActionData();
  const [form, fields] = useForm({
    lastResult: actionData?.result,
  });

  // you can access the message with `actionData?.message`
}
```

Any difference will it make with your suggestion on the route component side?

**Reply by adaboese** [reply]
The difference is that the convention is enforced at the library level, which makes it one less thing to think about. 

I already pass `form.context` to `FormProvider`.

```ts
const SignUpForm = () => {
  const fetcher = useFetcher<typeof action>();

  const [form, fields] = useForm({
    lastResult: fetcher.data?.submission,
    onValidate({ formData }) {
      return parseWithZod(formData, { schema: FormDataZodSchema });
    },
  });

  useEffect(() => {
    // We redirect user to this page after sign out, so we need to reset PostHog
    posthog.reset();
  }, []);

  const submitting = fetcher.state === 'submitting';

  return (
    <FormProvider context={form.context}>
      <fetcher.Form
        method="post"
        {...getFormProps(form)}
      >
        <Form.Layout>
          {fetcher.data?.message ? (
            <Form.TopMessage type={fetcher.data.message.type}>
              {fetcher.data.message.message}
            </Form.TopMessage>
          ) : null}
```

That is, in an ideal world I would not need:

```
         {fetcher.data?.message ? (
            <Form.TopMessage type={fetcher.data.message.type}>
              {fetcher.data.message.message}
            </Form.TopMessage>
          ) : null}
```

and could just derive if there is such a message from `form.context`.

```ts
export const TopMessage = () => {
  const context = useFormContext();

  if (!context.statusMessage) {
    return null;
  }

  return <div className={formMessage({ type: context.statusMessage.type })}>{context.statusMessage.message}</div>;
};
```
